### PR TITLE
Use shallow clone for git repositories by default (#129)

### PR DESCRIPTION
**Pull request by kingsamchen**
* use shallow clone for git repositories by default

* remove trailing spaces

### CODE REVIEW

File: cmake/CPM.cmake [hunk 1]
```diff
@@ -466,11 +480,25 @@ endfunction()
...
   else()
     string(REGEX MATCH "v?([0123456789.]*).*" _ ${GIT_TAG})
     SET(${RESULT} ${CMAKE_MATCH_1} PARENT_SCOPE)
   endif()
 endfunction()
+
+# guesses if the git tag is a commit hash or an actual tag or a branch nane.
+function(cpm_is_git_tag_commit_hash GIT_TAG RESULT)
+  string(LENGTH ${GIT_TAG} length)
+  if (NOT (length EQUAL 40))
```
[the remark this answers is not in the document]
Comment (kingsamchen) [reply cmake/CPM.cmake:495]: Thanks for the tip! and speaking of short hash, any length between `[7, 40)` could be a legal short hash, though 7-character is not enough for few super-big projects.

Should we treat lengths within the entire range as possible candidates of a commit hash ? given the guarantee of no commit hash should be mis-detected. 

Ref: https://stackoverflow.com/questions/18134627/how-much-of-a-git-sha-is-generally-considered-necessary-to-uniquely-identify-a

[the remark this answers is not in the document]
Comment (TheLartians) [reply cmake/CPM.cmake:495]: Yeah, that's a good point, I agree that we should guarantee that any legal git hash should continue working. And as most version tags would contain a special character or at least a 'v' prefix, we should still be getting shallow clones in almost all cases.

File: cmake/CPM.cmake [hunk 2]
```diff
@@ -230,6 +231,19 @@ function(CPMAddPackage)
...
+
+    # If GIT_SHALLOW is explicitly specified, hornor the value.
+    # Otherwise, enable shallow clone only when GIT_TAG is not a commit hash.
+    # Our guess may not be accurate, but it should guarantee no commit hash get mis-detected.
+    if (DEFINED CPM_ARGS_GIT_SHALLOW)
+      list(APPEND CPM_ARGS_UNPARSED_ARGUMENTS GIT_SHALLOW ${CPM_ARGS_GIT_SHALLOW})
+    else()
+      cpm_is_git_tag_commit_hash(${CPM_ARGS_GIT_TAG} IS_HASH)
+      if (NOT ${IS_HASH})
+        message(STATUS "GIT_TAG=${CPM_ARGS_GIT_TAG} is not a commit hash; enable shallow clone")
```
Comment (TheLartians) [inline cmake/CPM.cmake:243]: I wouldn't print internal config decisions to avoid flooding the output.

Comment (kingsamchen) [reply cmake/CPM.cmake]: Honestly, that was the first time I didn't use `CPM_SOURCE_CACHE`; and it indeed caused a checkout failure.

I noticed that without global cache, the name of source folder is always `foobar-src`, no version or some tag information entangled, which prevented re-downloading of the dependency; and the old folder is shallow-cloned, therefore resulted in checkout failure.

So I guess your suggestion is necessary to keep the local cache work.

Btw: the logging output is deleted.

File: cmake/CPM.cmake [hunk 3]
```diff
@@ -323,6 +310,18 @@ function(CPMAddPackage)
       list(APPEND CPM_ARGS_UNPARSED_ARGUMENTS DOWNLOAD_COMMAND "${CMAKE_COMMAND}")
       set(PACKAGE_INFO "${download_directory}")
     else()
+      # If GIT_SHALLOW is explicitly specified, hornor the value.
+      # Otherwise, enable shallow clone only when GIT_TAG is not a commit hash.
+      # Our guess may not be accurate, but it should guarantee no commit hash get mis-detected.
+      if (DEFINED CPM_ARGS_GIT_SHALLOW)
+        list(APPEND CPM_ARGS_UNPARSED_ARGUMENTS GIT_SHALLOW ${CPM_ARGS_GIT_SHALLOW})
+      else()
```
Comment (TheLartians) [inline cmake/CPM.cmake:318]: This part should still be considered in the default case (e.g. in line 233), as we still want to allow users to set `GIT_SHALLOW` explicitly in any case.

**Comment by TheLartians** [comment]
Looks great, thanks a lot for the PR! I'll test this in some large projects and will create a new release if everything works as expected.

**Pull request merged**